### PR TITLE
chore(ci): Centralize the bazel-diff call in a new job

### LINKED ISSUE #14553: Centralize tinder/bazel-diff call in bazel workflow
Depends on #14552 and #14536. The tinder/bazel-diff call should only happen once in the bazel workflow and the other jobs should depend on it.

### PR DESCRIPTION
**Pull request by LKreutzer**
Signed-off-by: Lars Kreutzer <lars.kreutzer@tngtech.com>

## Summary

- Resolves #14553 
- Bazel-diff is executed at most once and the result controls the jobs that depend on it.
- Bazel-diff is only used on PRs.
- Job flow diagram:
Screenshot from 2022-12-06 16-42-47
- Job schematic:
Screenshot from 2022-12-05 13-53-39
 

## Test Plan

- No changes: https://github.com/magma/magma/actions/runs/3628624756
- Changes to release, several python services, cc_test, go_test, py_test: https://github.com/magma/magma/actions/runs/3628677922
- Changes to release, python tests and one python service: https://github.com/magma/magma/actions/runs/3628329614/jobs/6119219562
- Changes to cc_test only (some jobs cancelled at the end due to push on the branch): https://github.com/magma/magma/actions/runs/3628767447
- Changes to python services (only main.py): https://github.com/magma/magma/actions/runs/3628876258
- Manual run on fork, no changes: https://github.com/LKreutzer/magma/actions/runs/3629554306

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: .github/workflows/bazel.yml [hunk 1]
```diff
@@ -353,49 +495,38 @@ jobs:
...
-
             printf '\r%s\r' '###############################' 1>&2
             printf '\r%s\r' 'Executing python import bazelification check.' 1>&2
             printf '\r%s\r' '###############################' 1>&2
             if [[ -n "${TEST_TARGETS}" ]];
             then
               if [[ "${TEST_TARGETS}" == "ALL_PYTHON_SERVICE_TARGETS" ]];
               then
-                bazel/scripts/test_python_service_imports.sh;
+                bazel/scripts/test_python_service_imports.sh 1>&2;
```
Comment (LKreutzer) [inline .github/workflows/bazel.yml:523]: These changes are unrelated, but this fixes errors in the ordering of the logging.

File: .github/workflows/bazel.yml [hunk 2]
```diff
@@ -85,7 +225,7 @@ jobs:
             bazel-target-rule: "cc_test"
           - bazel-config: "--config=production"
             bazel-target-rule: "cc_test"
-    name: Bazel Build & Test Job
+    name: Bazel Test Job
```
Comment (LKreutzer) [inline .github/workflows/bazel.yml:228]: Unrelated change: Makes it easier to read the job name (i.e. config type) because the name does not get cut off.

File: .github/workflows/bazel.yml [hunk 3]
```diff
@@ -67,14 +67,154 @@ jobs:
...
+              printf '\r%s\r' 'No relevant bazel targets are impacted, skipping all jobs.' 1>&2
+              printf "run_package_job=false" > /workspaces/magma/run_package_job.txt
+              printf "" > /workspaces/magma/all_unit_test_targets.txt
+              printf "" > /workspaces/magma/cc_unit_test_targets.txt
+              printf "" > /workspaces/magma/py_service_targets.txt
+              exit 0
+            fi
+
+            # Determine if the release is contained in the impacted targets list.
+            if [[ -z "$(grep '//lte/gateway/release:release_build' "$IMPACTED_TARGETS_FILE")" ]];
```
Comment (vktng) [inline .github/workflows/bazel.yml]: grep returns a different code if it does not find any results. You can achieve the same with:
`if ! grep -q '//lte/gateway/release:release_build' "$IMPACTED_TARGETS_FILE";`

Comment (LKreutzer) [reply .github/workflows/bazel.yml]: This does not work in bash I think because 
```
if [[ 0 ]];
then
    echo 'This is true'
else
    echo 'This is false'
fi

if [[ 1 ]];
then
    echo 'This is true'
else
    echo 'This is false'
fi
``` 
returns 
```
This is true
This is true
```
You then need to wrap it as a subshell in a string and compare to `0` or so. But that is essentially what it is doing now.

Comment (vktng) [reply .github/workflows/bazel.yml]: I did not wrap it inside [[ ]]
I'm just executing the command, then the if will follow the return code. The ! negates the effect.

Comment (vktng) [reply .github/workflows/bazel.yml]: You can change the line to what I wrote exactly.

Comment (LKreutzer) [reply .github/workflows/bazel.yml]: You are right, omitting the `[[ ]]` is key here. Pushed. I used the long option `--quiet` for better readability.

File: .github/workflows/bazel.yml [hunk 4]
```diff
@@ -133,19 +269,21 @@ jobs:
...
-
-              printf '\r%s\r' 'Determining impacted test targets ...' 1>&2
-              # Choose all impacted test targets of type 'matrix.bazel-target-rule' that are not tagged as manual.
-              TEST_TARGETS=$(bazel/scripts/filter_test_targets.sh "${{ matrix.bazel-target-rule }}" < "$IMPACTED_TARGETS_FILE")
+              if [[ "${{ matrix.bazel-target-rule }}" == '.*_test' ]];
+              then
+                TEST_TARGETS="${{ needs.bazel_diff.outputs.all_unit_test_targets }}"
+              else
+                TEST_TARGETS="${{ needs.bazel_diff.outputs.cc_unit_test_targets }}"
+                if [[ "$TEST_TARGETS" == "" ]];
```
Comment (vktng) [inline .github/workflows/bazel.yml]: Can we use a flag? e.g. -z or -n?

File: .github/workflows/bazel.yml [hunk 5]
```diff
@@ -260,8 +401,14 @@ jobs:
...
           echo "files_changed: $files_changed"
-
-          if [[ $bazel_build_and_test_success == "true" || $files_changed != "true" ]];
+          echo "bazel_diff_ran: $bazel_diff_ran"
+          echo "all_unit_test_targets: $all_unit_test_targets"
+
+          if [[ \
+              "$bazel_build_and_test_success" == "true" || \
+              "$files_changed" != "true" || \
+              ("$bazel_diff_ran" == "true" && "$all_unit_test_targets" == "") \
```
Comment (vktng) [inline .github/workflows/bazel.yml]: Same as above.

**Review by vktng — CHANGES_REQUESTED** [review]
Small changes requested, otherwise LGTM.

File: .github/workflows/bazel.yml (hunk 4 above)
Comment (LKreutzer) [reply .github/workflows/bazel.yml]: Pushed

File: .github/workflows/bazel.yml (hunk 5 above)
Comment (LKreutzer) [reply .github/workflows/bazel.yml]: Pushed

**Review by vktng — APPROVED** [review]
LGTM.
